File: src/brains/decision_tree.clj
(ns brains.decision-tree
  (require [clojure.set :as set]
           [clojure.contrib.string :as str]
           [brains.data :as data]
           [brains.probability-map :as pm]))

(defn- log2 [p]
  (/ (Math/log p) (Math/log 2)))

(defn- entropy [p]
  (- (* p (log2 p))))

(defn- igain
  "Calculates information gain on selecting this feature."
  [pm feature]
  (println feature)
  (let [outcome-e (for [o (pm/outcomes pm)]
                   (entropy (pm/p-o pm o)))
        all-e (for [outcome (pm/outcomes pm)
                    value (pm/feature-values pm feature)
                    :let [p-v (pm/p-f pm [feature value])
                          p (pm/p-of pm outcome [feature value])]
                    :when p]
                (* p-v (entropy p)))]
    (println all-e)
    (- (apply + outcome-e)
       (apply + all-e))))

(defn- feature-branch
  "Returns data points with matching feature->value, with that feature removed"
  [{:keys [data]} feature value]
  (for [d data :when (= value (feature d))]
    (dissoc d feature)))

(defn- build-tree
  "Builds decision tree from supplied data set and probability map."
  [data pm]
  (cond
   (= 1 (count (pm/outcomes pm)))
   (:leaf (first (pm/outcomes pm)))

   (or (empty? (:data data)) (empty? (pm/features pm)))
   (:leaf (pm/most-common-outcome pm))

   :else
   (let [gains (sort-by #(igain pm %) (pm/features pm))
         best-f (last gains)
         rest-f (butlast gains)
         children (for [v (pm/feature-values pm best-f)]
                    (let [new-points (feature-branch data best-f v)
                          new-data (assoc data :data new-points)]
                      [v, (build-tree new-data (pm/build new-data))]))]
     [:node best-f (into {} children)])))

(defn train [pm data])
(defn main [outcome datafile]
  (let [outcome (keyword outcome)
        data (data/load-from-file outcome datafile)
        points (shuffle (:data data))
        [test train] (split-at (/ (count points) 4)
                               points)
        pm (pm/build (assoc data :data train))]
    (doall
     (for [f (pm/features pm)]
       (println (igain pm f)))
     (println (build-tree data pm)))))

File: lib/brains/decision_tree.rb
module Brains

# A decision-tree classifier.
class DecisionTree

  def initialize(points)
    @most_common_class = nil
    @tree = train(points)
  end

  # Returns the class predicted for the given features.
  def classify(features)
    tree = @tree
    while true
      v = features[tree[0]]
      return @most_common_class unless v
      n = tree[1][v]
      if n.nil?
        return @most_common_class
      elsif n.class == String
        return n
      else
        tree = n
      end
    end
  end

  private
  def train(points)
    return nil if points.empty?

    pm = ProbabilityMap.new(points)
    if @most_common_class.nil?
      # save for later as default value
      @most_common_class = pm.most_common_class
    end
    return pm.classes.first if pm.classes.size == 1
    return nil if pm.features.empty?

    # find highest entropy feature to split with
    igains = pm.features.collect { |f| [info_gain(f, pm), f] }.sort.reverse
    best = igains.shift.last
    rest = igains.collect(&:last)
    tree = [best, {}]
    
    # split remaining data points on best feature
    pm.pf[best].keys.each do |val|
      newp = []
      points.each do |p|
        if p[1][best] == val
          p[1].delete(best)
          newp << p
        end
      end
      tree[1][val] = train(newp)
    end
    tree
  end

  def info_gain(feature, pm)
    class_e = pm.pc.keys.collect { |c| entropy(pm.pc[c] / pm.count) }
    all_e = []
    pm.pc.keys.each do |c|
      pm.pf.fetch(feature, {}).each do |v,pv|
        pv = sum(pv.values)
        p = pm.getpfc(feature, v, c)
        if p
          all_e << entropy(p / pm.count) * pv / pm.count
        end
      end
    end
    sum(class_e) - sum(all_e)
  end

  def sum(vals)
    vals.inject(0) { |a,b| a + b }
  end

  def entropy(p)
    -p * Math.log(p)
  end
end

end
